File: src/main/java/cn/motui/meican/service/DataServiceImpl.java
package cn.motui.meican.service;

import cn.motui.meican.MeiCanClient;
import cn.motui.meican.model.api.Account;
import cn.motui.meican.model.api.Address;
import cn.motui.meican.model.api.Dish;
import cn.motui.meican.model.api.vo.CalendarRestaurantVO;
import cn.motui.meican.model.api.vo.CalendarVO;
import cn.motui.meican.model.api.vo.CorpAddressVO;
import cn.motui.meican.model.api.vo.OrderDetailVO;
import cn.motui.meican.model.api.vo.RestaurantDishVO;
import cn.motui.meican.model.api.vo.RestaurantVO;
import cn.motui.meican.model.ui.DateData;
import cn.motui.meican.model.ui.OrderDetail;
import cn.motui.meican.model.ui.RestaurantData;
import com.google.common.collect.Lists;
import org.apache.commons.collections.CollectionUtils;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * 数据服务
 *
 * @author it.motui
 * @date 2021-01-30
 */
public class DataServiceImpl implements DataService {

  @Override
  public List<DateData> getDateData(LocalDateTime targetDateTime) {
    CalendarVO calendar = MeiCanClient.instance().calendar(targetDateTime, targetDateTime);
    List<DateData> dateDataList = Lists.newArrayList();
    if (CollectionUtils.isNotEmpty(calendar.getDateList())) {
      List<CalendarVO.CalendarItem> calendarItemList = calendar.getDateList().get(0).getCalendarItemList();
      if (CollectionUtils.isNotEmpty(calendarItemList)) {
        dateDataList = calendarItemList.stream()
            .map(CalendarVO.CalendarItem::toDateData)
            .collect(Collectors.toList());
      }
    }
    return dateDataList;
  }

  @Override
  public List<RestaurantData> getRestaurantData(String userTabUniqueId, LocalDateTime targetDateTime) {
    RestaurantVO restaurants = MeiCanClient.instance().restaurants(userTabUniqueId, targetDateTime);
    return restaurants.getRestaurantList().stream()
        .map(restaurant -> {
          // dish
          RestaurantDishVO restaurantDish = MeiCanClient.instance().restaurantDish(userTabUniqueId, targetDateTime,
              restaurant.getUniqueId());
          Set<Dish> dishSet = restaurantDish.getDishList().stream().filter(dish -> !dish.getSection())
              .collect(Collectors.toSet());
          return new RestaurantData(restaurant.getUniqueId(),
              restaurant.getName(),
              restaurant.getOpen(),
              Lists.newArrayList(dishSet)
          );
        }).collect(Collectors.toList());
  }

  @Override
  public List<Address> getAddress(String corpNamespace) {
    CorpAddressVO corpAddress = MeiCanClient.instance().address(corpNamespace);
    List<Address> addressList = Lists.newArrayList();
    if (CollectionUtils.isNotEmpty(corpAddress.getData().getAddressList())) {
      addressList = corpAddress.getData().getAddressList();
    }
    return addressList;
  }

  @Override
  public Account getAccount() {
    return MeiCanClient.instance().account();
  }

  @Override
  public OrderDetail getOrderDetail(String orderUniqueId) {
    OrderDetailVO orderDetailVO = MeiCanClient.instance().orderDetail(orderUniqueId);
    CalendarRestaurantVO calendarRestaurantVO = orderDetailVO.getCalendarRestaurants().get(0);
    CalendarRestaurantVO.DishItem dishItem = calendarRestaurantVO.getDishItemList().get(0);
    return new OrderDetail(
        orderDetailVO.getCorpOrderId(),
        orderDetailVO.getPickUpMessage(),
        orderDetailVO.getPostbox(),
        orderDetailVO.getProgressList(),
        calendarRestaurantVO.getUniqueId(),
        calendarRestaurantVO.getRestaurantName(),
        dishItem.getDish().toDish(),
        dishItem.getCount()
    );
  }

  @Override
  public String order(String userTabUniqueId, String addressUniqueId, LocalDateTime targetDateTime, Long dishId) {
    return MeiCanClient.instance().order(userTabUniqueId, addressUniqueId, targetDateTime, dishId);
  }
}

File: src/main/kotlin/cn/motui/meican/service/DataServiceImpl.kt
package cn.motui.meican.service

import cn.motui.meican.model.TabStatus
import cn.motui.meican.model.api.Account
import cn.motui.meican.model.api.Address
import cn.motui.meican.model.api.Dish
import cn.motui.meican.model.ui.OrderDetail
import cn.motui.meican.model.ui.RestaurantData
import cn.motui.meican.model.ui.TabData
import cn.motui.meican.util.meiCanClient
import cn.motui.meican.util.toLocalDateTime
import com.intellij.openapi.components.Service
import java.time.LocalDateTime

/**
 * 数据服务
 */
@Service
class DataServiceImpl : DataService {
    override fun getDateData(targetDateTime: LocalDateTime): MutableList<TabData> {
        val calendar = meiCanClient.calendar(targetDateTime, targetDateTime)
        val tabDataList: MutableList<TabData> = mutableListOf()
        calendar.dateList?.get(0)?.calendarItemList?.forEach {
            tabDataList.add(
                TabData(
                    it.title,
                    TabStatus.valueOf(it.status),
                    it.reason,
                    it.userTab.uniqueId,
                    it.userTab.corp.uniqueId,
                    it.userTab.corp.namespace,
                    it.corpOrderUser?.uniqueId,
                    it.targetTime.toLocalDateTime()
                )
            )
        }
        return tabDataList
    }

    override fun getRestaurantData(
        userTabUniqueId: String,
        targetDateTime: LocalDateTime
    ): MutableList<RestaurantData> {
        val restaurantDataList: MutableList<RestaurantData> = mutableListOf()
        val restaurant = meiCanClient.restaurants(userTabUniqueId, targetDateTime)
        restaurant.restaurantList?.forEach {
            val restaurantDish = meiCanClient.restaurantDish(userTabUniqueId, targetDateTime, it.uniqueId)
            val dishes: MutableList<Dish> = mutableListOf()
            restaurantDish?.dishList?.filter { item -> !item.section }
                ?.let { item -> dishes.addAll(item) }
            restaurantDataList.add(
                RestaurantData(
                    it.uniqueId,
                    it.name,
                    it.open,
                    dishes
                )
            )
        }
        return restaurantDataList
    }

    override fun getAddress(corpNamespace: String): List<Address> {
        val address = meiCanClient.address(corpNamespace)
        val addressList: MutableList<Address> = mutableListOf()
        address?.data?.addressList?.let { addressList.addAll(it) }
        return addressList
    }

    override fun refresh(username: String?, password: String?) {
        meiCanClient.refresh(username, password)
    }

    override fun getAccount(): Account {
        return meiCanClient.account()
    }

    override fun getOrderDetail(orderUniqueId: String): OrderDetail {
        val orderDetail = meiCanClient.orderDetail(orderUniqueId)
        val calendarRestaurant = orderDetail.calendarRestaurants[0]
        val dishItem = calendarRestaurant.dishItemList[0]
        return OrderDetail(
            orderDetail.uniqueId,
            orderDetail.corpOrderId,
            orderDetail.pickUpMessage,
            orderDetail.postbox,
            orderDetail.progressList,
            calendarRestaurant.uniqueId,
            calendarRestaurant.restaurantName,
            dishItem.dish.toDish(),
            dishItem.count,
            orderDetail.readyToDelete
        )
    }

    override fun order(
        userTabUniqueId: String,
        addressUniqueId: String,
        targetDateTime: LocalDateTime,
        dishId: Long
    ): String {
        return meiCanClient.order(userTabUniqueId, addressUniqueId, targetDateTime, dishId)
    }

    override fun cancelOrder(orderUniqueId: String) {
        meiCanClient.cancelOrder(orderUniqueId)
    }
}
